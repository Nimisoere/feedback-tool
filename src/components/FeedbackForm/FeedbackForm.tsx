import { yupResolver } from "@hookform/resolvers/yup";
import { useRouter } from "next/router";

import React from 'react'
import { useForm } from 'react-hook-form';
import * as yup from "yup";
import usePostFeedback from "../../hooks/api/usePostFeedback";
import RenderInput, { InputVariants } from './RenderInput';

const schema = yup.object({
  name: yup.string().required("Please enter your name"),
  email: yup.string().email().required("Please enter your email"),
  comment: yup.string().required("Please enter a comment"),
  rating: yup.number()
    .min(1, "Give a minimum rating of 1")
    .max(5, "Give a maximum rating of 5")
    .integer()
    .required("Please give a rating")
    .typeError("Rating must be a valid number")
});

interface FEEDBACK_INPUT {
  name: string;
  email: string;
  comment: string;
  rating: number
}

interface FEEDBACK_REQUEST {
  name: string;
  email: string;
  comment: string;
  rating: number;
  createdAt: Date;
}

const FeedbackForm: React.FC = () => {
  const router = useRouter()

  const { register, handleSubmit, formState } = useForm<FEEDBACK_INPUT>({
    resolver: yupResolver(schema),
  });
  const { loading, error, postFeedback } = usePostFeedback()

  const submit = (data: FEEDBACK_INPUT) => {
    postFeedback<FEEDBACK_REQUEST>({
      ...data,
      createdAt: new Date(),
    }, () => router.push("/view-feedback"))
  }

  return (
    <form onSubmit={handleSubmit(submit)}>
      {
        error && (
          <div className="alert-danger">
            {error?.message || "Unable to submit your feedback"}
          </div>
        )
      }
      <div className='flex flex-row gap-8 flex-wrap sm:flex-nowrap'>
        <div className='w-1/3'>
          <div className='flex flex-col'>
            <RenderInput
              id="name"
              label="Name"
              variant={InputVariants.INPUT}
              placeholder="Enter your name"
              error={formState.errors?.name}
              {...register("name")}
            />
            <RenderInput
              id="email"
              label="Email"
              variant={InputVariants.INPUT}
              placeholder="Enter your email"
              inputType="email"
              error={formState.errors?.email}
              {...register("email")}
            />
            <RenderInput
              id="rating"
              label="Rating"
              variant={InputVariants.STAR_RATING}
              placeholder="Enter your rating"
              error={formState.errors?.rating}
              {...register("rating")}
            />
          </div>
        </div>
        <div className='w-2/3'>
          <RenderInput
            id="comment"
            label="Comment"
            variant={InputVariants.TEXT_AREA}
            placeholder="Enter your comment"
            error={formState.errors?.comment}
            textareaProps={{
              rows: 10
            }}
            {...register("comment")}
          />
        </div>
      </div>
      <div className='flex justify-end'>
        <button disabled={loading} className='btn-primary' type='submit'>
          Submit{!!loading && "ing ..."}
        </button>
      </div>
    </form>
  )
}

export default FeedbackForm