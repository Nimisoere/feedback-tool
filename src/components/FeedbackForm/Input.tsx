import React from 'react'
import { range } from '../../utils/common';
import { InputVariants, RenderInputProps } from './RenderInput';

const Input: React.FC<Omit<RenderInputProps, "label">> = React.forwardRef<HTMLInputElement & HTMLTextAreaElement, Omit<RenderInputProps, "label">>(
  function getInput({ id, error, inputType, variant, textareaProps, rangeInputProps, ...props }, ref) {
    const inputclassname = `${error ? "border-red-300 focus:outline-red-300" : "border-slate-300 focus:outline-cyan-400"} bg-slate-50 border-2 rounded-md py-2 px-4 `
    switch (variant) {
      case InputVariants.INPUT:
        return (
          <input id={id} ref={ref} type={inputType || "text"} className={inputclassname} {...props} />
        )
      case InputVariants.TEXT_AREA:
        return (
          <textarea id={id} ref={ref} className={inputclassname} {...textareaProps} {...props} />
        )
      case InputVariants.STAR_RATING: {
        const min = rangeInputProps?.min || 1;
        const max = rangeInputProps?.max || 5
        const step = rangeInputProps?.step || 1

        return (
          <div className='flex flex-wrap sm:flex-col'>
            {
              range(min, max, step)?.map((count) => (
                <div className='text-slate-400 w-full' key={`range-${count}-${id}`}>
                  <input ref={ref} id={`range-${count}-${id}`} className="hidden peer" type="radio" {...props} value={count} />
                  <label className='cursor-pointer text-sm peer peer-checked:text-yellow-400' htmlFor={`range-${count}-${id}`}>
                    <span className='text-2xl'>&#11089;</span> ({count} {count === 1 ? "star" : "stars"})
                  </label>
                </div>
              ))
            }
          </div>
        )
      }
      default:
        return (
          <input id={id} ref={ref} type={inputType || "text"} className={inputclassname} {...props} />
        )
    }
  })

export default Input;