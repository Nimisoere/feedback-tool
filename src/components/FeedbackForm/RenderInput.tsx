import React from 'react'
import { FieldError, Merge } from 'react-hook-form';
import Input from './Input';

export enum InputVariants {
  INPUT = "input",
  TEXT_AREA = "text-area",
  STAR_RATING = "star-rating"
}

type RangeInputProps = {
  min: number
  max: number
  step: number
}

export type RenderInputProps = {
  id: string;
  label: string | React.ReactNode;
  variant: InputVariants
  inputType?: React.HTMLInputTypeAttribute
  placeholder?: string;
  error?: Merge<FieldError, (FieldError | undefined)[]>
  textareaProps?: React.TextareaHTMLAttributes<HTMLTextAreaElement>
  rangeInputProps?: RangeInputProps
  ref?: React.ForwardedRef<HTMLInputElement & HTMLTextAreaElement>
}

const RenderInput: React.FC<RenderInputProps> = React.forwardRef(function InputRenderer({ label, error, ...props }, ref) {
  return (
    <div className='flex flex-col mb-8'>
      <label className='font-medium text-sm'>{label}</label>
      <Input error={error} ref={ref} {...props} />
      {error && <span className="text-red-900 text-sm font-medium">{error?.message}</span>}
    </div>
  )
})

export default RenderInput

