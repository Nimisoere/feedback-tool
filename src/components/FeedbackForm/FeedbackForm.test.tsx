import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import mockRouter from 'next-router-mock';
import FeedbackForm from "./FeedbackForm";

jest.mock('next/router', () => require('next-router-mock'));

describe("Feedback form", () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl("/");
  });
  it("should match snapshot", async () => {
    const { getByText, container } = render(
      <FeedbackForm />
    );
    expect(container).toMatchSnapshot();
  });
});
