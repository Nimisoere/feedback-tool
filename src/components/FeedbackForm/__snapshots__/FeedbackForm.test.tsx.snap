// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback form should match snapshot 1`] = `
<div>
  <form>
    <div
      class="flex flex-row gap-8 flex-wrap sm:flex-nowrap"
    >
      <div
        class="w-full sm:w-1/3"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="flex flex-col mb-8"
          >
            <label
              class="font-medium text-sm"
            >
              Name
            </label>
            <input
              class="border-slate-300 focus:outline-cyan-400 bg-slate-50 border-2 rounded-md py-2 px-4 "
              id="name"
              name="name"
              placeholder="Enter your name"
              type="text"
            />
          </div>
          <div
            class="flex flex-col mb-8"
          >
            <label
              class="font-medium text-sm"
            >
              Email
            </label>
            <input
              class="border-slate-300 focus:outline-cyan-400 bg-slate-50 border-2 rounded-md py-2 px-4 "
              id="email"
              name="email"
              placeholder="Enter your email"
              type="email"
            />
          </div>
          <div
            class="flex flex-col mb-8"
          >
            <label
              class="font-medium text-sm"
            >
              Rating
            </label>
            <div
              class="flex flex-wrap sm:flex-col"
            >
              <div
                class="text-slate-400 w-full"
              >
                <input
                  class="hidden peer"
                  id="range-1-rating"
                  name="rating"
                  placeholder="Enter your rating"
                  type="radio"
                  value="1"
                />
                <label
                  class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
                  for="range-1-rating"
                >
                  <span
                    class="text-2xl"
                  >
                    ⭑
                  </span>
                   (
                  1
                   
                  star
                  )
                </label>
              </div>
              <div
                class="text-slate-400 w-full"
              >
                <input
                  class="hidden peer"
                  id="range-2-rating"
                  name="rating"
                  placeholder="Enter your rating"
                  type="radio"
                  value="2"
                />
                <label
                  class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
                  for="range-2-rating"
                >
                  <span
                    class="text-2xl"
                  >
                    ⭑
                  </span>
                   (
                  2
                   
                  stars
                  )
                </label>
              </div>
              <div
                class="text-slate-400 w-full"
              >
                <input
                  class="hidden peer"
                  id="range-3-rating"
                  name="rating"
                  placeholder="Enter your rating"
                  type="radio"
                  value="3"
                />
                <label
                  class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
                  for="range-3-rating"
                >
                  <span
                    class="text-2xl"
                  >
                    ⭑
                  </span>
                   (
                  3
                   
                  stars
                  )
                </label>
              </div>
              <div
                class="text-slate-400 w-full"
              >
                <input
                  class="hidden peer"
                  id="range-4-rating"
                  name="rating"
                  placeholder="Enter your rating"
                  type="radio"
                  value="4"
                />
                <label
                  class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
                  for="range-4-rating"
                >
                  <span
                    class="text-2xl"
                  >
                    ⭑
                  </span>
                   (
                  4
                   
                  stars
                  )
                </label>
              </div>
              <div
                class="text-slate-400 w-full"
              >
                <input
                  class="hidden peer"
                  id="range-5-rating"
                  name="rating"
                  placeholder="Enter your rating"
                  type="radio"
                  value="5"
                />
                <label
                  class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
                  for="range-5-rating"
                >
                  <span
                    class="text-2xl"
                  >
                    ⭑
                  </span>
                   (
                  5
                   
                  stars
                  )
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="w-full sm:w-2/3"
      >
        <div
          class="flex flex-col mb-8"
        >
          <label
            class="font-medium text-sm"
          >
            Comment
          </label>
          <textarea
            class="border-slate-300 focus:outline-cyan-400 bg-slate-50 border-2 rounded-md py-2 px-4 "
            id="comment"
            name="comment"
            placeholder="Enter your comment"
            rows="10"
          />
        </div>
      </div>
    </div>
    <div
      class="flex justify-end"
    >
      <button
        class="btn-primary w-full sm:w-auto"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
